File: documentation/code/qmatrix.cc
#include <iostream>
#include <typeinfo>

#include <exception>
#include <likelihood/qmatrix.h>
 

int main() {

  DCProgs::t_rmatrix matrix(5 ,5);
  matrix << -3050,        50,  3000,      0,    0, 
            2./3., -1502./3.,     0,    500,    0,  
               15,         0, -2065,     50, 2000,  
                0,     15000,  4000, -19000,    0,  
                0,         0,    10,      0,  -10;

  DCProgs::QMatrix qmatrix(matrix, /*nopen=*/2);
 
  std::cout << qmatrix << std::endl;

  if(qmatrix.nopen != 2) return 1;
  if(qmatrix.nshut() != 3) return 1;
  if( ((qmatrix.matrix - matrix).array().abs() > 1e-12).any() ) return 1;
  
  // Compute sum over rows, row by row.
  for(DCProgs::t_int i(0); i < qmatrix.matrix.rows(); ++i) 
    std::cout << "sum(row[" << i << "]): " << qmatrix.matrix.row(i).sum() << std::endl;

  // Compute sum over rows, but let eigen do it.
  std::cout << "Sum over rows: " << qmatrix.matrix.rowwise().sum().transpose() << std::endl;
  
  // Check that sum over rows returns zero.
  if( (qmatrix.matrix.rowwise().sum().array().abs() > 1e-10).any() ) throw std::exception();
  
  // Gets AA block.
  std::cout << "AA block:\n---------\n" << qmatrix.aa() << std::endl;
                
  // aa returns a wrapper around the matrix block. Hence memory is the same.
  if(&(qmatrix.aa()(0,0)) != &(qmatrix(0, 0))) throw std::exception();
 
  return 0;
}

File: documentation/code/qmatrix.py
from numpy import sum, abs, all
from dcprogs.likelihood import QMatrix

matrix = [ [-3050,        50,  3000,      0,    0], 
           [2./3., -1502./3.,     0,    500,    0], 
           [   15,         0, -2065,     50, 2000], 
           [    0,     15000,  4000, -19000,    0], 
           [    0,         0,    10,      0,  -10] ]

qmatrix = QMatrix(matrix, 2);
 
print(qmatrix)

assert qmatrix.nopen == 2
assert qmatrix.nshut == 3
assert all(abs(qmatrix.matrix - matrix) < 1e-12)

  
# Compute sum over rows, row by row.
for i, row in enumerate(qmatrix.matrix): print("sum(row[{0}]) = {1}".format(i, sum(row)))

# Compute sum over rows, but let numpy do it.
print("Sum over rows: {0}".format(sum(qmatrix.matrix, axis=1)))
  
# Check that sum over rows returns zero.
assert all(abs(sum(qmatrix.matrix, axis=1)) < 1e-12)
  
# Gets AA block.
print("AA block:\n---------\n{0}".format(qmatrix.aa))
                
# aa is simply wrapper around the matrix block 
qmatrix.aa[0, 0] = 42
assert abs(qmatrix[0, 0] - 42) < 1e-12
